### PR TITLE
auth, etcdserver: add a method for updating backend during apply snap…

### PR DESCRIPTION
**Pull request by mitake**
…shot

This commit adds a new method UpdateBackend() to auth.AuthStore for
updating backend during apply snapshot. It follows a manner of the
lessor.

### CODE REVIEW

**Comment by philips** [comment]
@mitake Can you explain a bit more on what you mean by: "It follows a manner of the lessor."? I don't think I follow how this relates to a lease.

File: auth/store.go [hunk 1]
```diff
@@ -33,16 +33,19 @@ var (
 type AuthStore interface {
 	// AuthEnable() turns on the authentication feature
 	AuthEnable()
+
+	// UpdateBackend updates authStore.be during EtcdServer.applySnapshot()
+	UpdateBackend(b backend.Backend)
```
Comment (xiang90) [inline auth/store.go]: probably Recover? Some recovering needs to be done, like reading the auth info from new backend to rebuild the in memory state.

File: auth/store.go [hunk 2]
```diff
@@ -52,15 +55,18 @@ func (as *authStore) AuthEnable() {
 	plog.Noticef("Authentication enabled")
 }
 
-func NewAuthStore(bgetter backendGetter) *authStore {
-	b := bgetter.Backend()
-	tx := b.BatchTx()
+func (as *authStore) UpdateBackend(be backend.Backend) {
+	as.be = be
```
Comment (xiang90) [inline auth/store.go]: todo: add recovery logic.

File: auth/store.go [hunk 3]
```diff
@@ -33,16 +33,19 @@ var (
...
 	// AuthEnable() turns on the authentication feature
 	AuthEnable()
+
+	// UpdateBackend updates authStore.be during EtcdServer.applySnapshot()
+	UpdateBackend(b backend.Backend)
 }
 
 type authStore struct {
-	bgetter backendGetter
+	be backend.Backend
```
Comment (xiang90) [inline auth/store.go]: remove backendGetter interface at line 22?

**Comment by mitake** [comment]
@philips this PR is based on this comment: https://github.com/coreos/etcd/pull/4714#issuecomment-196141930

etcdserver passes a new backend to lease during applying snapshot. Unlike lease, current auth requires etcdserver in its constructor and calls `Backend()` when it needs backend. I think the way of lease is cleaner so creating this PR. There's no intention of connecting auth and lease in this change.

File: auth/store.go (hunk 1 above)
Comment (mitake) [inline auth/store.go]: Yes, I misunderstood about the interface of lease. The recovery process is required.

File: auth/store.go (hunk 3 above)
Comment (mitake) [inline auth/store.go]: Oops, thanks for your pointing.

**Comment by mitake** [comment]
@xiang90 updated for your comments, PTAL.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**